SlicerLabkey integration

andrejstuden opened the pull request:
Adding a pair of extension which together comprise integration of LabKey database access to Slicer. LabKey (http://labkey.com) is a database management program which can integrate images as well. Slicer is used as image viewing and processing toolbox.

- [x ] Extension has a reasonable name (not too general, not too narrow, suggests what the extension is for)
- [ x] Repository name is Slicer+ExtensionName
- [ x] Extension description summarizes in 1-2 sentences what the extension is usable (should be understandable for non-experts)
- [x ] Extension URL and revision (scmurl, scmrevision) is correct, consider using a branch name (master, release, ...) instead of a specific git has to avoid re-submitting pull request whenever the extension is updated
- [ x] Extension icon URL is correct
- [x ] Screenshot URLs (screenshoturls) are correct, contains at least one
- [ x] Homepage URL points to valid webpage containing the following:
  - [ x] Extension name
  - [x ] Short description: 1-2 sentences, which summarizes what the extension is usable for
  - [ x] At least one nice, informative image, that illustrates what the extension can do. It may be a screenshot.
  - [x ] Description of contained modules: at one sentence for each module
  - [x ] Tutorial: step-by-step description of at least the most typical use case, include a few screenshots, provide download links to sample input data set
  - [ ] Publication: link to publication and/or to PubMed reference (if available)
  - [x ] License: We suggest you use a permissive license that includes patent and contribution clauses.  This will help protect developers and ensure the code remains freely available.  We suggest you use the [Slicer License](https://github.com/Slicer/Slicer/blob/master/License.txt) or the [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0). Always mention in your README file the license you have chosen.  If you choose a different license, explain why to the extension maintainers. Depending on the license we may not be able to host your work.
Read [here](https://opensource.guide/legal/#which-open-source-license-is-appropriate-for-my-project) to learn more about licenses.

pieper commented:
Sounds interesting!  But I can't find any information about what the extension actually does.  I'm not familiar with Labkey.  Can you add more description?

andrejstuden commented:
I know, I am pretty bad that way. Where should I put the description - the project page at gitlab, readme.md or a different place?

pieper commented:
I would think the gitlab readme would be great - it doesn't need to be a lot of info, just enough to help people get a sense of the essential functionality and use.

BTW, I'm not able to see the code or clone the repository - maybe you need to make it public?  I am able to see the gitlab page but there's no code visible.

```
$(cd /tmp/; git clone git://wiscigt.powertheword.com/andrej.studen/SlicerLabkeyDICOMtools)
Cloning into 'SlicerLabkeyDICOMtools'...
fatal: unable to connect to wiscigt.powertheword.com:
wiscigt.powertheword.com[0: 168.61.213.220]: errno=Operation timed out
```

andrejstuden commented:
Updated the readme in SlicerLabkeyDICOMtools. Changed settings for the repository for public access. While project access was enabled, the code access was limited to members. Rectified that, should be OK now.

pieper commented:
Thanks - the readme helps.

I'm still getting the timeout on the git clone though.

lassoan commented:
I'm trying to understand from [README.md](http://wiscigt.powertheword.com/andrej.studen/SlicerLabkeyDICOMtools/blob/SlicerExtensionIndex/README.md) what this extension does. Could you please clarify:
- what do you mean by "targeted DICOM parsing in concert with SlicerLabkeyPythonAPI"?
- "downloading DICOM files" - from where? how?

"The extension consists of four classes, mostly API with minimal GUI": The extension must contain Slicer modules. You can include Python modules in subdirectories in one or more modules but they should be placed in a subdirectory (see for example [here](https://github.com/Slicer/Slicer/tree/master/Modules/Scripted/DataProbe/DataProbeLib)).

importDicom is a modified clone of Slicer's DICOM import infrastructure. If you find that some things are inadequate then please submit a pull request with suggested modifications to the existing functions.

exportDicom is similarly a modified Slicer core feature and instead of introducing and maintaining a new variant we should update the existing infrastructure to fulfill your needs.

loadDicom: similarly to the other modules, it is unclear for me how it is different from existing core features and why it is duplicated&modified instead of improved in Slicer core.

vtkInterface: I think all these features are already available in slicer.util and sitkUtils, but if we missed anything or the API should be improved then send a pull request.

Overall, I think most changes should be applied to the Slicer core and you could add 1-2 modules with a Python library and a user-friendly GUI to implement some LabKey-specific features.

andrejstuden commented:
Hello!

This requires a bit more thought from my end. Just to let you know it might take a while, as I have to go back and trace the issues I had with the original Slicer code. 

Thanks!

Andrej

lassoan commented:
Also note that the DICOM browser has been completely reworked in most recent Slicer Preview Releases. The DICOM browser is now separated from the database, so you can import and load data more cleanly, without relying on GUI widgets. You can also run DICOM indexing in the background, without blocking the user interface.

andrejstuden commented:
Hi!

Your comments spurred a much needed housekeeping exercise. Now DICOMtools is part of SlicerLabkeyExtension as its most important function is to rely on remote database and to open and convert remote DICOM files to nodes. Any rework of Slicer code that removes dependency on database is welcome for that part, as in this case the database is remote rather than local. To make it clearer, more screenshots were made and the description got more precise.
Where the software access is concerned - read/write git@ access requires user privileges. Read only access runs through http and not https. If that is a strict requirement, I'll deal with it. Everything else should be fine.

Andrej

pieper commented on SlicerLabkeyExtension.s4ext:
```diff
@@ -0,0 +1,44 @@
+
+# This is source code manager (i.e. svn)
+scm git
+scmurl git://wiscigt.powertheword.com/andrej.studen/SlicerLabkeyExtension.git
```
This URL times out.  You should change to this, which works:

http://wiscigt.powertheword.com/andrej.studen/SlicerLabkeyExtension.git

pieper commented on SlicerLabkeyPythonAPI.s4ext:
```diff
@@ -0,0 +1,44 @@
+#
```
It looks like this one should be removed?

andrejstuden commented:
Thanks, updated both. Let me know if that looks better.

Andrej

pieper commented:
I haven't been able to look too closely but it looks good from what I can see.

lassoan commented:
@andrejstuden I see the recent update. Let us know when you want us to review/merge your pull request.

lassoan commented:
This pull request still has not been merged. Is there still an interest to make this extension show up in the Extensions manager in Slicer?

lassoan commented:
I close this extension to make all active pull requests more clearly visible. @andrejstuden please reopen anytime if you are still interested in submitting this extension to the extension index.
